src/Features/CSharp/Portable/ForeachToFor/CSharpForeachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,151 @@
+            // easily reused here.
+            return foreachInfo.ForEachStatement.Parent.IsKind(SyntaxKind.Block);
+        }
+
+        protected override (SyntaxNode start, SyntaxNode end) GetForEachBody(ForEachStatementSyntax foreachStatement)
+        {
+            return (foreachStatement.Statement, foreachStatement.Statement);
+        }
+
+        protected override void ConvertToForStatement(SemanticModel model, ForEachInfo foreachInfo, SyntaxEditor editor, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var generator = editor.Generator;
+            var foreachStatement = foreachInfo.ForEachStatement;
+
+            var foreachCollectionExpression = foreachStatement.Expression;
+            var collectionVariable = GetCollectionVariableName(model, generator, foreachInfo, foreachCollectionExpression);
+
+            // first, see whether we need to introduce new statement to capture collection
+            IntroduceCollectionStatement(model, foreachInfo, editor, foreachCollectionExpression, collectionVariable);
+
+            // create new index varialbe name
+            var indexString = CreateUniqueName(model, foreachStatement.Statement, "i");
+
+            // put variable statement in body
+            var bodyStatement = GetForLoopBody(generator, foreachInfo, collectionVariable, indexString);
+
+            // create for statement from foreach statement
+            var forStatement = SyntaxFactory.ForStatement(
+                SyntaxFactory.VariableDeclaration(
+                    SyntaxFactory.IdentifierName("var"),
+                    SyntaxFactory.SingletonSeparatedList(
+                        SyntaxFactory.VariableDeclarator(
+                            generator.Identifier(indexString).WithAdditionalAnnotations(RenameAnnotation.Create()),
+                            argumentList: default,
+                            SyntaxFactory.EqualsValueClause((ExpressionSyntax)generator.LiteralExpression(0))))),
+                SyntaxFactory.SeparatedList<ExpressionSyntax>(),
+                (ExpressionSyntax)generator.LessThanExpression(
+                    generator.IdentifierName(indexString),
+                    generator.MemberAccessExpression(collectionVariable, foreachInfo.CountName)),
+                SyntaxFactory.SingletonSeparatedList<ExpressionSyntax>(
+                    SyntaxFactory.PostfixUnaryExpression(
+                        SyntaxKind.PostIncrementExpression, SyntaxFactory.IdentifierName(indexString))),
+                bodyStatement);
+
+            // let leading and trailing trivia set
+            if (!foreachInfo.RequireCollectionStatement)
+            {
+                forStatement = forStatement.WithLeadingTrivia(foreachStatement.GetLeadingTrivia());
+            }
+
+            forStatement = forStatement.WithCloseParenToken(foreachStatement.CloseParenToken);
```

Comment:
you shoudl keep both the open and close parens of the foreach statement.  You can provide them in situ when creating the ForStatement above.